### DISCUSSION TITLE
Overland Maps

### DISCUSSION THREAD

**Discussion by windu-ant** [GoMudEngine/GoMud#36] [Ideas]
The in game map is wonderful so far. However, one of my favorite things on other codebases is the support for overland maps to occupy that space between towns and castles. Instead of builders having to create a "filler space" like a forest with little actual stuff in it or one single long path to make it feel spacious, the overland can take this over and mostly be an ASCII map with basic descriptions. 

I believe a lot of MUD's have adapted the code from AFKMud to handle overland maps, and the beauty of that system is it uses a PNG image, say 1000x1000 pixels, and each pixel represents a room if it is that corresponding RGB color.

The overland actually exists as only one (maybe a few?) vnum's but uses a grid numbering system for where the player is. You can mapedit in game to change the overland map, or edit a PNG with an editor (paint, gimp, etc), or a generator.

Code reference:
https://github.com/Arthmoor/AFKMud/blob/master/src/overland.cpp
https://github.com/Arthmoor/AFKMud/blob/master/areadocs/RGB.txt

You can use something like Azgaar's Fantasy Map generator and just modify the biome RGB colors to match what the MUD wants and export your map as a PNG and add it to the map directory, along with a few references in code. Quite literally creating an overland map within minutes. I even successfully managed this on my first try.

Overland image generator:
https://azgaar.github.io/Fantasy-Map-Generator/

**Comment by Volte6** [comment]
I like the idea of it... but I don't fully understand.

Is this intended to replace the ascii maps I have (for example), or to be a larger scale map that mostly deals with landmarks? I could think of a few ways something like this could work.

**Reply by windu-ant** [reply]
It wouldn't be a replacement but instead a supplement to the current maps. Something like the old overland maps in Final Fantasy or other RPG's. Once you touch a landmark tile (#) on the overland map it changes to the regular maps.

![image](https://github.com/Volte6/GoMud/assets/20514384/60cc67b4-93c1-47f4-b007-395da669e8be)
Once you touch a # tile then you get the "regular" map and descrips:
![image](https://github.com/Volte6/GoMud/assets/20514384/9f1ddc0f-1149-4799-9db9-94a721af3b06)

**Reply by Volte6** [reply]
So traversal is more in the style of a Roguelike when moving in the overland mode?

**Reply by windu-ant** [reply]
When you're on the world map it can be quite rogue-like I suppose. In that picture above the player is the red @, and two mobs are the purple @'s. The system even supports random encounters on the map. Once you go into a town or area it turns into the bottom example, much more like the maps in GoMUD.

**Comment by Volte6** [comment]
Well I admit I come from a generation of MUD playing that predates this, so it's a bit foreign to me... but overall I think it's a really nice feature. I think aspirationally I could see adding this down the line as an optional feature... I do feel like it would really benefit from the alternative screen buffer mode and redrawing elements on screen, but that would likely conflict greatly with mudlet, which refuses to respect moving the cursor around... so that's an interesting problem.

overall i like it though, and will keep this around for future reference.

**Reply by windu-ant** [reply]
I think the only way for mudlet to get around this would be to use something like GMCP within mudlet to send data in a separate channel and then Mudlet could display it within a separate window. Some Muds have set up some pretty snazzy mudlet interfaces. Midnight sun comes to mind: https://midnightsun2.org/clients.html

I think since a lot of clients don't support the clear screen or cursor movement features a lot of us have just gotten used to the map being redrawn every time you move, and I don't find it too jarring.

**Reply by Volte6** [reply]
Properly detecting and supporting mudlet in particular will need to be a feature addition.

Or possibly, I will introduce a player config that allows disabling cursor move redrawn, to more widely support the behavior, and detect/default specific clients.

I like the overworked concept, if nothing else as an optional feature to enable. I can think of some interesting features and puzzles that can be represented in that format as well, such as flying/sailing into the unknown, or sokoban puzzles.

If I'm honest this won't be implemented (by me) for a while until I really wrap my head around how I want to do it, and address other issues i havent settled (magic for example) but I'm sold on the concept.
